File: src/main/java/com/ptaf/ai/index/YamlKeyIndex.java
package com.ptaf.ai.index;

import org.yaml.snakeyaml.Yaml;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Indexes YAML keys across configured framework folders and flattens nested keys into dot notation.
 */
public final class YamlKeyIndex {
    private final Set<String> normalizedKeys;
    private final Map<String, String> keySourceType;

    public YamlKeyIndex(Set<String> normalizedKeys, Map<String, String> keySourceType) {
        this.normalizedKeys = Set.copyOf(normalizedKeys);
        this.keySourceType = Map.copyOf(keySourceType);
    }

    public Set<String> normalizedKeys() {
        return normalizedKeys;
    }

    public Map<String, String> keySourceType() {
        return keySourceType;
    }

    public static YamlKeyIndex build(Path projectRoot, Map<String, String> yamlPaths) throws IOException {
        Set<String> keys = new LinkedHashSet<>();
        Map<String, String> sourceType = new LinkedHashMap<>();
        Yaml yaml = new Yaml();

        for (Map.Entry<String, String> entry : yamlPaths.entrySet()) {
            String namespace = entry.getKey();
            Path root = projectRoot.resolve(entry.getValue()).normalize();
            if (!Files.isDirectory(root)) {
                continue;
            }
            for (Path file : listYamlFiles(root)) {
                String text = Files.readString(file, StandardCharsets.UTF_8);
                Object parsed = yaml.load(text);
                flatten(namespace, "", parsed, keys, sourceType);
            }
        }

        return new YamlKeyIndex(keys, sourceType);
    }

    @SuppressWarnings("unchecked")
    private static void flatten(
            String namespace,
            String currentPath,
            Object value,
            Set<String> keys,
            Map<String, String> sourceType
    ) {
        if (value instanceof Map<?, ?> map) {
            for (Map.Entry<?, ?> entry : map.entrySet()) {
                String key = String.valueOf(entry.getKey());
                String nextPath = currentPath.isEmpty() ? key : currentPath + "." + key;
                String fullKey = normalizeKey(namespace + "." + nextPath);
                keys.add(fullKey);
                sourceType.putIfAbsent(fullKey, namespace);
                flatten(namespace, nextPath, entry.getValue(), keys, sourceType);
            }
            return;
        }
        if (value instanceof List<?> list) {
            for (int i = 0; i < list.size(); i++) {
                String nextPath = currentPath + "." + i;
                String fullKey = normalizeKey(namespace + "." + nextPath);
                keys.add(fullKey);
                sourceType.putIfAbsent(fullKey, namespace);
                flatten(namespace, nextPath, list.get(i), keys, sourceType);
            }
        }
    }

    public static String normalizeKey(String raw) {
        if (raw == null) {
            return "";
        }
        String normalized = raw.trim().toLowerCase();
        normalized = normalized.replaceAll("\\.+", ".");
        normalized = normalized.replaceAll("^\\.+", "");
        normalized = normalized.replaceAll("\\.+$", "");
        return normalized;
    }

    private static List<Path> listYamlFiles(Path root) throws IOException {
        try (Stream<Path> walk = Files.walk(root)) {
            return walk
                    .filter(Files::isRegularFile)
                    .filter(path -> {
                        String n = path.toString().toLowerCase();
                        return n.endsWith(".yml") || n.endsWith(".yaml");
                    })
                    .sorted(Comparator.comparing(Path::toString))
                    .collect(Collectors.toCollection(ArrayList::new));
        }
    }
}

File: src/main/ptaf/ai/index/yaml_key_index.py
"""Indexes YAML keys across configured framework folders."""

from __future__ import annotations

from pathlib import Path
from typing import Any

import yaml


class YamlKeyIndex:
    def __init__(self, normalized_keys: set[str], key_source_type: dict[str, str]) -> None:
        self._normalized_keys = set(normalized_keys)
        self._key_source_type = dict(key_source_type)

    def normalized_keys(self) -> set[str]:
        return self._normalized_keys

    def key_source_type(self) -> dict[str, str]:
        return self._key_source_type

    @classmethod
    def build(cls, project_root: Path, yaml_paths: dict[str, str]) -> YamlKeyIndex:
        keys: set[str] = set()
        source_type: dict[str, str] = {}
        for namespace, relative_path in yaml_paths.items():
            root = (project_root / relative_path).resolve()
            if not root.is_dir():
                continue
            for file_path in sorted(root.rglob("*")):
                if not file_path.is_file():
                    continue
                name = file_path.name.lower()
                if not (name.endswith(".yml") or name.endswith(".yaml")):
                    continue
                parsed = yaml.safe_load(file_path.read_text(encoding="utf-8"))
                cls._flatten(namespace, "", parsed, keys, source_type)
        return cls(keys, source_type)

    @staticmethod
    def normalize_key(raw: str | None) -> str:
        if raw is None:
            return ""
        normalized = raw.strip().lower()
        while ".." in normalized:
            normalized = normalized.replace("..", ".")
        normalized = normalized.strip(".")
        return normalized

    @classmethod
    def _flatten(
        cls,
        namespace: str,
        current_path: str,
        value: Any,
        keys: set[str],
        source_type: dict[str, str],
    ) -> None:
        if isinstance(value, dict):
            for key, child in value.items():
                next_path = key if not current_path else f"{current_path}.{key}"
                full_key = cls.normalize_key(f"{namespace}.{next_path}")
                keys.add(full_key)
                source_type.setdefault(full_key, namespace)
                cls._flatten(namespace, next_path, child, keys, source_type)
            return
        if isinstance(value, list):
            for index, child in enumerate(value):
                next_path = f"{current_path}.{index}"
                full_key = cls.normalize_key(f"{namespace}.{next_path}")
                keys.add(full_key)
                source_type.setdefault(full_key, namespace)
                cls._flatten(namespace, next_path, child, keys, source_type)
